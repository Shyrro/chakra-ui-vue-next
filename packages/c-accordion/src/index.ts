import { h, defineComponent, PropType } from "vue"
import { chakra, DOMElements } from "@chakra-ui/vue-system"

const CAccordion = defineComponent({
  props: {
    as: {
      type: [String] as PropType<DOMElements>,
      default: "div",
    },
  },
  setup(props, { slots, attrs }) {
    return () =>
      h(chakra(props.as, { label: "accordion" }), { ...attrs }, slots)
  },
})

export default CAccordion
