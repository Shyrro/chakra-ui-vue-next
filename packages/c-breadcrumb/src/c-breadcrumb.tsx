/**
 * Hey! Welcome to @chakra-ui/vue-next CBreadcrumb
 *
 * Breadcrumbs help users visualize their current location in relation to the rest of the website or application by showing the hierarchy of pages
 *
 * @see Docs     https://next.vue.chakra-ui.com/breadcrumb
 * @see Source   https://github.com/chakra-ui/chakra-ui-vue-next/blob/master/packages/c-breadcrumb/src/c-breadcrumb/c-breadcrumb.ts
 * @see WAI-ARIA https://www.w3.org/TR/wai-aria-practices-1.2
 */

import { h, defineComponent, PropType, ConcreteComponent, Component, computed, cloneVNode, VNode, VNodeProps, mergeProps } from 'vue'
import {
  chakra,
  HTMLChakraProps,
  SystemProps,
  ThemingProps,
  StylesProvider,
  useMultiStyleConfig,
  useStyles,
  SystemStyleObject
} from '@chakra-ui/vue-system'
import { filterUndefined } from '@chakra-ui/utils'
import { getValidChildren, SNA, SNAO } from '@chakra-ui/vue-utils'

/**
 * CBreadcrumb (root)
 */


export interface BreadcrumbOptions {
  /**
   * The visual separator between each breadcrumb item
   * @type string | ConcreteComponent | Component
   */
  separator?: string | ConcreteComponent | Component
  /**
   * The left and right margin applied to the separator
   * @type SystemProps["mx"]
   */
  spacing?: SystemProps['mx']
}

export interface BreadcrumbProps
  extends HTMLChakraProps<'nav'>,
    BreadcrumbOptions,
    ThemingProps<'Breadcrumb'> {}


/**
 * CBreadcrumb is used to render a breadcrumb navigation landmark.
 * It renders a `nav` element with `aria-label` set to `breadcrumb`
 *
 * @see Docs https://next.vue.chakra-ui.com/breadcrumb
 */
export const CBreadcrumb = defineComponent(
  (props: BreadcrumbProps, { attrs, slots }) => {
    const themingProps = computed<ThemingProps>(() =>
      filterUndefined({
        colorScheme: props.colorScheme,
        variant: props.variant,
        size: props.size,
        styleConfig: props.styleConfig,
      })
    )

    const styles = useMultiStyleConfig('Breadcrumb', themingProps.value)
    StylesProvider(styles)

    const separator = computed(() => slots?.separator?.() || props.separator)

    return () => {
      const validChildren = getValidChildren(slots)
      const count = validChildren.length

      const children = validChildren.map((vnode: VNode<unknown, unknown, BreadcrumbOptions>, index: number) => cloneVNode(vnode, {
        separator: separator.value,
        spacing: props.spacing,
        isLastChild: count === index + 1,
      }))

      return (
        <chakra.nav
          as={props.as}
          aria-label="breadcrumb"
          __css={styles.value.container}
          {...attrs}
        >
          <chakra.ol __label="chakra-breadcrumb__list">
            {() => children}
          </chakra.ol>
        </chakra.nav>
      )
    }
  }
)

// @ts-ignore "name" property is typically read-only for functional components
CBreadcrumb.name = 'CBreadcrumb'
CBreadcrumb.props = {
  separator: {
    type: SNAO as PropType<BreadcrumbOptions['separator']>,
    default: '/'
  },
  spacing: {
    type: SNA as PropType<BreadcrumbOptions['spacing']>,
    default: '0.5rem'
  },
  as: {
    type: [String, Object] as PropType<DOMElements | Component | string>,
    default: 'nav',
  },
}


/**
 * CBreadcrumbSeparator
 */

export interface BreadcrumbSeparatorProps extends HTMLChakraProps<"div"> {
  /**
   * @type SystemProps["mx"]
   */
  spacing?: SystemProps["mx"]
}

/**
 * The `CBreadcrumbSeparator` component is the separator for
 * each breacrumb item.
 */
export const CBreadcrumbSeparator = defineComponent((props: BreadcrumbSeparatorProps, { attrs, slots }) => {
  const styles = useStyles()
  const separatorStyles = computed<SystemStyleObject>(() => ({
    display: 'flex',
    mx: props.spacing,
    ...styles.value.separator,
  }))

  return () => (
    <chakra.span
      role="presentation"
      __label="chakra-breadcrumb__separator"
      {...attrs}
      __css={separatorStyles.value}
    >
      {slots}
    </chakra.span>
  )
})

CBreadcrumbSeparator.props = {
  spacing: CBreadcrumb.props.spacing,
}


// @ts-ignore "name" property is typically read-only for functional components
CBreadcrumbSeparator.name = 'CBreadcrumbSeparator'


/**
 * CBreadcrumbItem
 */

interface BreadcrumbItemOptions extends BreadcrumbOptions {
  isCurrentPage?: boolean
  isLastChild?: boolean
}

export interface BreadcrumbItemProps
  extends BreadcrumbItemOptions,
    HTMLChakraProps<"li"> {}

export const CBreadcrumbItem = defineComponent((props: BreadcrumbItemProps, { attrs, slots }) => {
  const styles = useStyles()
  const itemStyles = computed<SystemStyleObject>(() => ({
    display: "inline-flex",
    alignItems: "center",
    ...styles.value.item,
  }))

  return () => {
    const validChildren = getValidChildren(slots)
    const children = validChildren.map((vnode: VNode<unknown, unknown, BreadcrumbItemOptions>) => {
      // @ts-expect-error The "name" property is not typed on `VNodeTypes` but we need to access it during runtime
      if (vnode.type.name === 'CBreadcrumbLink') {
        return cloneVNode(vnode, {
          isCurrentPage: props.isCurrentPage,
        })
      }

      // @ts-expect-error The "name" property is not typed on `VNodeTypes` but we need to access it during runtime
      if (vnode.type.name === 'CBreadcrumbSeparator') {
        return cloneVNode(vnode, {
          spacing: props.spacing,
          children: vnode.children || { default: () => props.separator },
        })
      }

      return vnode
    })

    return (
      <chakra.li __label="chakra-breadcrumb__list-it" __css={itemStyles.value}>
        {children}
        {!props.isLastChild && (
          // @ts-expect-error
          <CBreadcrumbSeparator spacing={props.spacing}>
            {() => props.separator}
          </CBreadcrumbSeparator>
        )}
      </chakra.li>
    )
  }
})

// @ts-ignore "name" property is typically read-only for functional components
CBreadcrumbItem.name = 'CBreadcrumbItem'
CBreadcrumbItem.props = {
  ...CBreadcrumb.props,
  isLastChild: Boolean as PropType<boolean>,
  isCurrentPage: Boolean as PropType<boolean>,
}

/**
 * CBreadcrumbLink
 */

export interface BreadcrumbLinkProps extends HTMLChakraProps<"a"> {
  isCurrentPage?: boolean
}

/**
 * BreadcrumbLink link.
 *
 * It renders a `span` when it matches the current link. Otherwise,
 * it renders an anchor tag.
 */
export const CBreadcrumbLink = defineComponent((props: BreadcrumbLinkProps, { attrs, slots }) => {
  const styles = useStyles()
  
  return () => {
    if (props.isCurrentPage) {
      return (
        <chakra.span __label="chakra-breadcrumb__link" aria-current="page" __css={styles.value.link} as={props.as} {...attrs}>
          {slots}
        </chakra.span>
      )
    }

    return (
      <chakra.a __label="chakra-breadcrumb__link" as={props.as} __css={styles.value.link} {...attrs}>
        {slots}
      </chakra.a>
    )
  }
})

// @ts-ignore "name" property is typically read-only for functional components
CBreadcrumbLink.name = 'CBreadcrumbLink'
CBreadcrumbLink.props = {
  isCurrentPage: Boolean as PropType<boolean>,
}