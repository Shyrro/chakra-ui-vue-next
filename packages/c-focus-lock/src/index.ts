export * from './c-focus-lock'
