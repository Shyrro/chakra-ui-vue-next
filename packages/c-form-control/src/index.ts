export * from './c-form-control'
export * from './use-form-control'
export * from './c-form-error'