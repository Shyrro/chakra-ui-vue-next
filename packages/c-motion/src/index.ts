export * from './c-motion'
export * from './animate-presence'
export * from './motion-utils'
