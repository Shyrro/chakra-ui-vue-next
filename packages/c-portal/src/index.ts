export { default as CPortal } from "./portal"
