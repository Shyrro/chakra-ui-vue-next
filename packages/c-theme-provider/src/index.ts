import {
  h,
  defineComponent,
  Fragment,
  PropType,
  provide,
  inject,
  computed,
} from "vue"
import { Theme } from "@chakra-ui/vue-theme"

const CThemeProvider = defineComponent({
  name: "CThemeProvider",
  props: {
    value: {
      type: [Object] as PropType<Theme>,
      default: () => undefined,
    },
  },
  setup(props, { slots }) {
    const pluginTheme = inject("$chakraTheme")
    const applicationTheme = computed(() => props.value || pluginTheme)
    provide("$chakraTheme", applicationTheme.value)
    return () => h(Fragment, slots.default?.({ $chakraTheme: props.value }))
  },
})

export default CThemeProvider
