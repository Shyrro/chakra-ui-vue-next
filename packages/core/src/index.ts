import { computed, Plugin, Ref, ref, watch } from 'vue'
import defaultTheme, { ColorMode } from '@chakra-ui/vue-theme'
import { toCSSVar, WithCSSVar } from '@chakra-ui/styled-system'
import { chakra, injectGlobal, css } from '@chakra-ui/vue-system'
import internalIcons from './icon.internals'
import { extendTheme, ThemeOverride } from './extend-theme'
import { MergedIcons, parseIcons } from './parse-icons'
import { mode } from '@chakra-ui/vue-theme-tools'

interface ExtendIconsPath {
  path: string
  viewBox?: string
}

interface IconsOptions {
  pack?: 'fa' | 'fe'
  library?: {}
  extend?: Record<string, ExtendIconsPath>
}
export interface ChakraUIVuePluginOptions {
  cssReset?: boolean
  extendTheme?: ThemeOverride
  icons?: IconsOptions
  defaultColorMode?: ColorMode
}

const ChakraUIVuePlugin: Plugin = {
  install(app, options: ChakraUIVuePluginOptions = { cssReset: true }) {
    // 1. Get theme value
    // 2. Parse theme tokens to CSS variables
    // 3. Inject all CSS variables as theme object
    const theme = options.extendTheme || defaultTheme
    const computedTheme = computed<WithCSSVar<ThemeOverride>>(() =>
      toCSSVar(theme)
    )
    injectGlobal({
      ':root': computedTheme.value.__cssVars,
    })

    let libraryIcons = options.icons?.library || {}
    let extendedIcons = options.icons?.extend || {}

    // Initialize colormode
    const colorMode: ColorMode = theme.config?.initialColorMode || 'light'

    // Bind theme to application global properties and provide to application
    app.config.globalProperties.$chakraTheme = computedTheme.value
    app.provide('$chakraTheme', computedTheme.value as ThemeOverride)

    // Provide initial colormode
    app.config.globalProperties.$initialColorMode = colorMode
    const colorModeRef = ref<ColorMode>(colorMode)
    app.provide<Ref<ColorMode>>('$chakraColorMode', colorModeRef)

    libraryIcons = parseIcons(libraryIcons)

    // Merge internal icons and library icons
    const mergedIcons: MergedIcons = {
      ...internalIcons,
      ...libraryIcons,
      ...extendedIcons,
    }

    app.provide('$chakraIcons', mergedIcons)

    // Set color mode property
    app.config.globalProperties.$mode = mode
  },
}

export interface ThemeProviderProps extends ThemeOverride {}
export default ChakraUIVuePlugin
export { extendTheme }

// Export chakra factory function
export { chakra }

/**
 *
 * Component exports
 * ==================
 *
 * Dear contributors,
 *
 * Please keep these exports in Alphabetical order :)
 */

// A
export * from '@chakra-ui/c-accordion'
export * from '@chakra-ui/c-alert'

// B
export * from '@chakra-ui/c-button'

// C
export * from '@chakra-ui/c-color-mode'
export * from '@chakra-ui/c-close-button'
export * from '@chakra-ui/c-code'

// F
export * from '@chakra-ui/c-flex'
export * from '@chakra-ui/c-focus-lock'

// I
export * from '@chakra-ui/c-icon'

// L
export * from '@chakra-ui/vue-layout'

// M
export * from '@chakra-ui/c-modal'
export * from '@chakra-ui/c-motion'

// P
export * from '@chakra-ui/c-popper'
export * from '@chakra-ui/c-portal'

// R
export * from '@chakra-ui/c-reset'

// S
export * from '@chakra-ui/c-spinner'
export * from '@chakra-ui/c-scroll-lock'

// T
export * from '@chakra-ui/c-theme-provider'

// V
export * from '@chakra-ui/c-visually-hidden'

// OTHERS
export * from '@chakra-ui/vue-composables'
export * from '@chakra-ui/vue-a11y'

/**
 *
 * Directives exports
 * ==================
 *
 * Dear contributors,
 *
 * Please keep these exports in Alphabetical order :)
 */
