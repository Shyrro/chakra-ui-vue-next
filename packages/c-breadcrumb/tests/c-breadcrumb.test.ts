import { render } from '../../test-utils/src'
import { CBreadcrumb } from '../src'

it('should render properly', () => {
  const { asFragment } = render(CBreadcrumb)
  expect(asFragment()).toMatchSnapshot()
})