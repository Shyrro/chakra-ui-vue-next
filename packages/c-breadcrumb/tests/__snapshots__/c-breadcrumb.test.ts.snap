// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly 1`] = `
<DocumentFragment>
  <nav
    aria-label="breadcrumb"
    class="css-0"
  >
    <ol
      class="chakra-chakra-breadcrumb__list css-0"
    />
  </nav>
</DocumentFragment>
`;
